### PR TITLE
Implementation of WebView for new SantaGUI

### PR DESCRIPTION
**Pull request by DOHtero**

### CODE REVIEW

**Comment by DOHtero** [comment]
Google has approved my membership to push updates

**Comment by DOHtero** [comment]
I signed it!

**Comment by DOHtero** [comment]
I thought I did all the steps to get this taken care of. Any advice on what I can do while I am away?

**Comment by russellhancox** [comment]
Sorry Daniel, we've been a bit swamped, I'll try and get this reviewed in the next few days.

**Comment by DOHtero** [comment]
No worries Russ. I totally understand with everything that is going on internally.

File: Source/SantaGUI/SNTMessageWindowController.h [hunk 1]
```diff
@@ -23,13 +24,18 @@
...
+
 
 - (instancetype)initWithEvent:(SNTStoredEvent *)event andMessage:(NSString *)message;
 
 - (IBAction)showWindow:(id)sender;
 - (IBAction)closeWindow:(id)sender;
-- (IBAction)showCertInfo:(id)sender;
+- (IBAction)showCertInfo;
 
 ///  Reference to the "Bundle Hash" label in the XIB. Used to remove if application
```
Comment (russellhancox) [inline Source/SantaGUI/SNTMessageWindowController.h]: Remove the remaining IBOutlet properties in this header (`bundleHashLabel`, `bundleHashTitle`, `hashingIndicator`, `openEventButton`) as they presumably aren't used anymore.

File: Source/SantaGUI/SNTMessageWindowController.m [hunk 2]
```diff
@@ -57,6 +57,38 @@ - (instancetype)initWithEvent:(SNTStoredEvent *)event andMessage:(NSString *)mes
   return self;
```
Comment (russellhancox) [inline Source/SantaGUI/SNTMessageWindowController.m]: Remove the `applicationNameLabel` property, I assume it isn't used.

File: Source/SantaGUI/SNTMessageWindowController.m [hunk 3]
```diff
@@ -57,6 +57,38 @@ - (instancetype)initWithEvent:(SNTStoredEvent *)event andMessage:(NSString *)mes
...
+  [self.webView setDrawsBackground:NO];
+  self.window.movableByWindowBackground = YES;
+  self.webView.shouldCloseWithWindow = YES;
+  
+  // Turn off scrollbars in the frame
+  [[[self.webView mainFrame] frameView] setAllowsScrolling:NO];
+  
+  // Load the page
+  [self.webView.windowScriptObject setValue:self forKey:@"AppDelegate"];
+  [self.mainWindow.contentView addSubview:_webView];
```
Comment (russellhancox) [inline Source/SantaGUI/SNTMessageWindowController.m]: I'm assuming `mainWindow` can be replaced with `window` here.

File: Source/SantaGUI/SNTMessageWindowController.m [hunk 4]
```diff
@@ -57,6 +57,38 @@ - (instancetype)initWithEvent:(SNTStoredEvent *)event andMessage:(NSString *)mes
   return self;
 }
 
+- (void)windowDidLoad {
+  [self.window setFrame:NSMakeRect(self.window.frame.origin.x, self.window.frame.origin.y, 0, 0)
+                display:NO];
+  self.webView.frame = NSMakeRect(0, 0, 0, 0);
+  [self.webView setDrawsBackground:NO];
```
Comment (russellhancox) [inline Source/SantaGUI/SNTMessageWindowController.m]: Could this use the property setter?

`self.webView.drawsBackground = NO`

File: Source/SantaGUI/SNTMessageWindowController.m [hunk 5]
```diff
@@ -109,24 +159,24 @@ - (void)loadWindow {
 
 - (IBAction)showWindow:(id)sender {
   [(SNTMessageWindow *)self.window fadeIn:sender];
+  [self addObserver:self forKeyPath:@"self.event.fileBundleHash" options:0 context:NULL];
 }
 
-- (IBAction)closeWindow:(id)sender {
+- (void)closeWindow:(id)sender {
```
Comment (russellhancox) [inline Source/SantaGUI/SNTMessageWindowController.m]: As this is no longer used from any Cocoa controls (and you're making it no longer an IBAction) we might as well override the `close` method from NSWindowController. Change the method signature to:

`- (void)close`

File: Source/SantaGUI/SNTMessageWindowController.h [hunk 6]
```diff
@@ -23,13 +24,18 @@
...
+@property(weak) IBOutlet NSWindow *mainWindow;
+@property(weak) IBOutlet WebView *webView;
+
 
 - (instancetype)initWithEvent:(SNTStoredEvent *)event andMessage:(NSString *)message;
 
 - (IBAction)showWindow:(id)sender;
 - (IBAction)closeWindow:(id)sender;
-- (IBAction)showCertInfo:(id)sender;
+- (IBAction)showCertInfo;
```
Comment (russellhancox) [inline Source/SantaGUI/SNTMessageWindowController.h]: Remove these 3 methods - the first is defined in NSWindowController, the second is being changed (see my comment in the implementation) and the last is only used within the class.

File: Source/SantaGUI/SNTMessageWindowController.m [hunk 7]
```diff
@@ -171,9 +221,91 @@ - (NSString *)publisherInfo {
   }
 }
 
-- (NSAttributedString *)attributedCustomMessage {
-  return [SNTBlockMessage attributedBlockMessageForEvent:self.event
-                                           customMessage:self.customMessage];
+- (NSString *)attributedCustomMessage {
+  return [SNTBlockMessage blockMessageForGUI:self.event
```
Comment (russellhancox) [inline Source/SantaGUI/SNTMessageWindowController.m]: I believe this will fit on one line.

File: Source/SantaGUI/SNTMessageWindowController.m [hunk 8]
```diff
@@ -171,9 +221,91 @@ - (NSString *)publisherInfo {
...
+  }
+  if (self.event.parentName) {
+    dataSet[@"parent"] = self.event.parentName;
+  }
+  if (self.event.ppid) {
+    dataSet[@"ppid"] = self.event.ppid;
+  }
+  if (self.event.executingUser) {
+    dataSet[@"user"] = self.event.executingUser;
+  }
```
Comment (russellhancox) [inline Source/SantaGUI/SNTMessageWindowController.m]: Most of these could be single line if statements to save space, e.g:

```objc
if (self.attributedCustomMessage) dataSet[@"message"] = self.attributedCustomMessage;
if (self.event.fileBundleName) dataSet[@"application"] = self.event.fileBundleName;
```

File: Source/SantaGUI/SNTMessageWindowController.m [hunk 9]
```diff
@@ -171,9 +221,91 @@ - (NSString *)publisherInfo {
...
+  NSURLRequest *URL = [NSURLRequest requestWithURL:[NSURL fileURLWithPath:htmlPath]];
+  return URL;
+}
+
+// Allows Javascript to only call the methods below
++ (BOOL)isSelectorExcludedFromWebScript:(SEL)selector {
+  if (selector == @selector(santaData)
+      || selector == @selector(showCertInfo)
+      || selector == @selector(acceptFunction)
+      || selector == @selector(ignoreFunction:)) {
```
Comment (russellhancox) [inline Source/SantaGUI/SNTMessageWindowController.m]: Keep the ORs with the left condition, e.g.:

```objc
  if (selector == @selector(santaData) ||
      selector == @selector(showCertInfo) ||
      selector == @selector(acceptFunction) ||
      selector == @selector(ignoreFunction:)) {
```

File: Source/SantaGUI/SNTMessageWindowController.m [hunk 10]
```diff
@@ -171,9 +221,91 @@ - (NSString *)publisherInfo {
...
+  }
+  
+  NSData *dataTransform = [NSJSONSerialization dataWithJSONObject:dataSet options:0 error:NULL];
+  NSString *json = [[NSString alloc] initWithData:dataTransform encoding:NSUTF8StringEncoding];
+  return json;
+}
+
+// Allows links in WebView to open in default Web browser instead of within the WebView
+- (void)webView:(WebView *)webView
+decidePolicyForNavigationAction:(NSDictionary *)actionInformation
```
Comment (russellhancox) [inline Source/SantaGUI/SNTMessageWindowController.m]: Indent this line by 4 spaces, then colon-align the arguments after this one.

File: Source/SantaGUI/SNTNotificationManager.m [hunk 11]
```diff
@@ -100,27 +100,27 @@ - (void)postClientModeNotification:(SNTClientMode)clientmode {
   NSString *customMsg;
   switch (clientmode) {
     case SNTClientModeMonitor:
-      un.informativeText = @"Switching into Monitor mode";
-      customMsg = [[SNTConfigurator configurator] modeNotificationMonitor];
-      customMsg = [SNTBlockMessage stringFromHTML:customMsg];
-      if (customMsg.length) un.informativeText = customMsg;
-      break;
+    un.informativeText = @"Switching into Monitor mode";
```
Comment (russellhancox) [inline Source/SantaGUI/SNTNotificationManager.m]: Indent the body of each case by 2 spaces (as it was before)

File: Source/SantaGUI/SNTNotificationManager.m [hunk 12]
```diff
@@ -100,27 +100,27 @@ - (void)postClientModeNotification:(SNTClientMode)clientmode {
...
+    return;
   }
   [[NSUserNotificationCenter defaultUserNotificationCenter] deliverNotification:un];
 }
 
 - (void)postBlockNotification:(SNTStoredEvent *)event withCustomMessage:(NSString *)message {
   // See if this binary is already in the list of pending notifications.
   NSPredicate *predicate =
-      [NSPredicate predicateWithFormat:@"event.fileSHA256==%@", event.fileSHA256];
+  [NSPredicate predicateWithFormat:@"event.fileSHA256==%@", event.fileSHA256];
```
Comment (russellhancox) [inline Source/SantaGUI/SNTNotificationManager.m]: Undo this change, this line should be indented.

File: Source/SantaGUI/SNTNotificationManager.m [hunk 13]
```diff
@@ -149,7 +149,7 @@ - (void)postBlockNotification:(SNTStoredEvent *)event withCustomMessage:(NSStrin
   // This includes making windows.
   dispatch_async(dispatch_get_main_queue(), ^{
     SNTMessageWindowController *pendingMsg =
-        [[SNTMessageWindowController alloc] initWithEvent:event andMessage:message];
+    [[SNTMessageWindowController alloc] initWithEvent:event andMessage:message];
```
Comment (russellhancox) [inline Source/SantaGUI/SNTNotificationManager.m]: Same

File: Source/common/SNTBlockMessage.m [hunk 14]
```diff
@@ -39,28 +48,17 @@ + (NSAttributedString *)attributedBlockMessageForEvent:(SNTStoredEvent *)event
     message = [[SNTConfigurator configurator] unknownBlockMessage];
     if (!message) {
       message = @"The following application has been blocked from executing<br />"
-                @"because its trustworthiness cannot be determined.";
+      @"because its trustworthiness cannot be determined.";
```
Comment (russellhancox) [inline Source/common/SNTBlockMessage.m]: Undo

File: Source/common/SNTBlockMessage.m [hunk 15]
```diff
@@ -39,28 +48,17 @@ + (NSAttributedString *)attributedBlockMessageForEvent:(SNTStoredEvent *)event
...
     }
   } else {
     message = [[SNTConfigurator configurator] bannedBlockMessage];
     if (!message) {
       message = @"The following application has been blocked from executing<br />"
-                @"because it has been deemed malicious.";
+      @"because it has been deemed malicious.";
```
Comment (russellhancox) [inline Source/common/SNTBlockMessage.m]: Undo

Comment (russellhancox) [inline Source/common/SNTBlockMessage.m]: Undo

File: Source/SantaGUI/SNTMessageWindowController.h [hunk 16]
```diff
@@ -23,13 +24,18 @@
 ///
 ///  Controller for a single message window.
 ///
-@interface SNTMessageWindowController : NSWindowController
+@interface SNTMessageWindowController : NSWindowController<NSApplicationDelegate,
+                                                           WebPolicyDelegate,
+                                                           WebFrameLoadDelegate>
+@property(weak) IBOutlet NSWindow *mainWindow;
```
Comment (DOHtero) [inline Source/SantaGUI/SNTMessageWindowController.h]: ack

File: Source/SantaGUI/SNTMessageWindowController.h [hunk 17]
```diff
@@ -23,13 +24,18 @@
...
+@property(weak) IBOutlet WebView *webView;
```
Comment (DOHtero) [inline Source/SantaGUI/SNTMessageWindowController.h]: Doing this gives me a Property not found error

File: Source/SantaGUI/SNTMessageWindowController.m (hunk 9 above)
Comment (DOHtero) [inline Source/SantaGUI/SNTMessageWindowController.m]: ack

File: Source/SantaGUI/SNTMessageWindowController.m (hunk 8 above)
Comment (DOHtero) [inline Source/SantaGUI/SNTMessageWindowController.m]: ack

File: Source/SantaGUI/SNTMessageWindowController.m (hunk 10 above)
Comment (DOHtero) [inline Source/SantaGUI/SNTMessageWindowController.m]: ack

File: Source/SantaGUI/SNTNotificationManager.m (hunk 11 above)
Comment (DOHtero) [inline Source/SantaGUI/SNTNotificationManager.m]: ack

File: Source/SantaGUI/SNTNotificationManager.m (hunk 12 above)
Comment (DOHtero) [inline Source/SantaGUI/SNTNotificationManager.m]: ack

File: Source/SantaGUI/SNTNotificationManager.m (hunk 13 above)
Comment (DOHtero) [inline Source/SantaGUI/SNTNotificationManager.m]: ack

File: Source/common/SNTBlockMessage.m (hunk 14 above)
Comment (DOHtero) [inline Source/common/SNTBlockMessage.m]: ack

File: Source/common/SNTBlockMessage.m (hunk 15 above)
Comment (DOHtero) [inline Source/common/SNTBlockMessage.m]: ack

Comment (DOHtero) [inline Source/common/SNTBlockMessage.m]: Same as above

File: Source/SantaGUI/SNTMessageWindowController.m (hunk 7 above)
Comment (DOHtero) [inline Source/SantaGUI/SNTMessageWindowController.m]: It does. ack

File: Source/SantaGUI/SNTMessageWindowController.m (hunk 2 above)
Comment (DOHtero) [inline Source/SantaGUI/SNTMessageWindowController.m]: ApplicationNameLabel is being used in line 156

File: Source/SantaGUI/SNTMessageWindowController.m (hunk 4 above)
Comment (DOHtero) [inline Source/SantaGUI/SNTMessageWindowController.m]: ack

File: Source/SantaGUI/SNTMessageWindowController.m (hunk 5 above)
Comment (DOHtero) [inline Source/SantaGUI/SNTMessageWindowController.m]: closeWindow is being using throughout the entire SNTMessageWindowcontroller. When I change it to what you suggestion and trigger a SantaBlock, I am unable to click the buttons

File: Source/SantaGUI/SNTMessageWindowController.h (hunk 16 above)
Comment (DOHtero) [inline Source/SantaGUI/SNTMessageWindowController.h]: ack

File: Source/SantaGUI/SNTMessageWindowController.h (hunk 1 above)
Comment (DOHtero) [inline Source/SantaGUI/SNTMessageWindowController.h]: ack

File: Source/SantaGUI/SNTMessageWindowController.h (hunk 17 above)
Comment (DOHtero) [inline Source/SantaGUI/SNTMessageWindowController.h]: ack

File: Source/SantaGUI/SNTMessageWindowController.m (hunk 2 above)
Comment (DOHtero) [inline Source/SantaGUI/SNTMessageWindowController.m]: ack

File: Source/SantaGUI/SNTMessageWindowController.m (hunk 3 above)
Comment (DOHtero) [inline Source/SantaGUI/SNTMessageWindowController.m]: ack
